cmake/embed_clang.cmake

```diff
@@ -0,0 +1,140 @@
+if(EMBED_CLANG)
+  include(ExternalProject)
+  include(ProcessorCount)
+
+  if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+    set(EMBEDDED_BUILD_TYPE "RelWithDebInfo")
+  elseif(CMAKE_BUILD_TYPE STREQUAL "Release")
+    set(EMBEDDED_BUILD_TYPE "MinSizeRel")
+  else()
+    set(EMBEDDED_BUILD_TYPE ${CMAKE_BUILD_TYPE})
+  endif()
+
+  if(${LLVM_VERSION} VERSION_EQUAL "9" OR ${LLVM_VERSION} VERSION_GREATER "9" )
+    set(LLVM_FULL_VERSION "9.0.1")
+    set(CLANG_DOWNLOAD_URL "https://github.com/llvm/llvm-project/releases/download/llvmorg-${LLVM_FULL_VERSION}/clang-${LLVM_FULL_VERSION}.src.tar.xz")
+    set(CLANG_URL_CHECKSUM "SHA256=5778512b2e065c204010f88777d44b95250671103e434f9dc7363ab2e3804253")
+  elseif(${LLVM_VERSION} VERSION_EQUAL "8" OR ${LLVM_VERSION} VERSION_GREATER "8" )
+    set(LLVM_FULL_VERSION "8.0.1")
+    set(CLANG_DOWNLOAD_URL "https://github.com/llvm/llvm-project/releases/download/llvmorg-${LLVM_FULL_VERSION}/cfe-${LLVM_FULL_VERSION}.src.tar.xz")
+    set(CLANG_URL_CHECKSUM "SHA256=70effd69f7a8ab249f66b0a68aba8b08af52aa2ab710dfb8a0fba102685b1646")
+  elseif(${LLVM_VERSION} VERSION_EQUAL "7" OR ${LLVM_VERSION} VERSION_GREATER "7" )
+    set(LLVM_FULL_VERSION "7.1.0")
+    set(CLANG_DOWNLOAD_URL "https://github.com/llvm/llvm-project/releases/download/llvmorg-${LLVM_FULL_VERSION}/cfe-${LLVM_FULL_VERSION}.src.tar.xz")
+    set(CLANG_URL_CHECKSUM "SHA256=e97dc472aae52197a4d5e0185eb8f9e04d7575d2dc2b12194ddc768e0f8a846d")
+  else()
+    message(FATAL_ERROR "No supported LLVM version has been specified with LLVM_VERSION (LLVM_VERSION=${LLVM_VERSION}), aborting")
+  endif()
+
+  set(CLANG_PATCH_COMMAND "/bin/true")
+  if(NOT EMBED_LLVM)
+    # If not linking and building against embedded LLVM, patches may need to
+    # be applied to work with the distribution LLVM. This is handled by a
+    # helper function
+    include(embed_helpers)
+    prepare_clang_patches(CLANG_PATCH_COMMAND)
+  endif()
+
+  # TODO dalehamel
+  # is there a way to inherit the number of jobs from a cmake variable?
```

Comment:
It does work, I had to move away from this approach though to be able to specify the make command to just build one target, "libclang".

And for android, I want to specify just the libraries I want to build - this requires overriding the make command.

There may be a CMake generator expression, since this is a dynamic value passed to make / ninja.

Reply:
Overall I want to use CMake to control the build where I can, as this breaks using ninja as the generator unnecessarily. For now it is more flexible for prototyping.

An easy way around this is to have an if/else to define the external project using the the default BUILD_COMMAND, and one that specifies it if the variable isn't empty. I don't think Cmake has a more elegant way to do this, perhaps I am wrong?

Some of the 3rd party libs needed to make the android build work don't use cmake, so the pattern of writing a build command is already necessary. 

I agree it can and should be improved. I'll aim to go for the if/else approach for now, unless you have any other ideas?

Reply:
> An easy way around this is to have an if/else to define the external project using the the default BUILD_COMMAND, and one that specifies it if the variable isn't empty. I don't think Cmake has a more elegant way to do this, perhaps I am wrong?

Not sure. I only know enough cmake to get by :). I figure it's best to get it working and if we run into issues we can tweak/hack as necessary.

Reply:
I have changed the code to only override BUILD_COMMAND this variable is set.

Reply:
I found a much better approach!

This can all be done with interpolation, see:

https://chromium.googlesource.com/external/github.com/llvm-mirror/llvm/+/refs/heads/master/cmake/modules/LLVMExternalProjectUtils.cmake#168

and https://chromium.googlesource.com/external/github.com/llvm-mirror/llvm/+/refs/heads/master/cmake/modules/LLVMExternalProjectUtils.cmake#224

I'll fix up my branch to use this.

Reply:
done